Flaky TestSchedulerProcessor_processQueriesOnSingleStream

Seen a flaky run of `TestSchedulerProcessor_processQueriesOnSingleStream` ([CI](https://github.com/grafana/mimir/actions/runs/8659078973/job/23744238911?pr=7882)):

```
--- FAIL: TestSchedulerProcessor_processQueriesOnSingleStream (1.65s)
    --- FAIL: TestSchedulerProcessor_processQueriesOnSingleStream/should_not_cancel_query_execution_if_scheduler_client_returns_a_non-cancellation_error (0.51s)
        scheduler_processor_test.go:253: 
            	Error Trace:	/__w/mimir/mimir/pkg/querier/worker/scheduler_processor_test.go:253
            	Error:      	Not equal: 
            	            	expected: 2
            	            	actual  : 1
            	Test:       	TestSchedulerProcessor_processQueriesOnSingleStream/should_not_cancel_query_execution_if_scheduler_client_returns_a_non-cancellation_error
            	Messages:   	Expected number of calls (2) does not match the actual number of calls (1).
FAIL
FAIL	github.com/grafana/mimir/pkg/querier/worker	4.198s
```

Another example in https://github.com/grafana/mimir/actions/runs/9909349624/job/27377221017?pr=8685:
```
--- FAIL: TestSchedulerProcessor_processQueriesOnSingleStream (1.54s)
    --- FAIL: TestSchedulerProcessor_processQueriesOnSingleStream/should_not_cancel_query_execution_if_scheduler_client_returns_a_non-cancellation_error (0.52s)
        scheduler_processor_test.go:255: 
            	Error Trace:	/__w/mimir/mimir/pkg/querier/worker/scheduler_processor_test.go:255
            	Error:      	Not equal: 
            	            	expected: 2
            	            	actual  : 1
            	Test:       	TestSchedulerProcessor_processQueriesOnSingleStream/should_not_cancel_query_execution_if_scheduler_client_returns_a_non-cancellation_error
            	Messages:   	Expected number of calls (2) does not match the actual number of calls (1).
```

another instance https://github.com/grafana/mimir/actions/runs/10955420076/job/30419274402?pr=9344

```
--- FAIL: TestSchedulerProcessor_processQueriesOnSingleStream (1.64s)
    --- FAIL: TestSchedulerProcessor_processQueriesOnSingleStream/should_not_cancel_query_execution_if_scheduler_client_returns_a_non-cancellation_error (0.51s)
        scheduler_processor_test.go:255: 
            	Error Trace:	/__w/mimir/mimir/pkg/querier/worker/scheduler_processor_test.go:255
            	Error:      	Not equal: 
            	            	expected: 2
            	            	actual  : 1
            	Test:       	TestSchedulerProcessor_processQueriesOnSingleStream/should_not_cancel_query_execution_if_scheduler_client_returns_a_non-cancellation_error
            	Messages:   	Expected number of calls (2) does not match the actual number of calls (1).
FAIL
```

another instance https://github.com/grafana/mimir/actions/runs/11040649844/job/30669127966?pr=9413

```
 --- FAIL: TestSchedulerProcessor_processQueriesOnSingleStream (1.56s)
    --- FAIL: TestSchedulerProcessor_processQueriesOnSingleStream/should_not_cancel_query_execution_if_scheduler_client_returns_a_non-cancellation_error (0.51s)
        scheduler_processor_test.go:255: 
            	Error Trace:	/__w/mimir/mimir/pkg/querier/worker/scheduler_processor_test.go:255
            	Error:      	Not equal: 
            	            	expected: 2
            	            	actual  : 1
            	Test:       	TestSchedulerProcessor_processQueriesOnSingleStream/should_not_cancel_query_execution_if_scheduler_client_returns_a_non-cancellation_error
            	Messages:   	Expected number of calls (2) does not match the actual number of calls (1).
FAIL
FAIL	github.com/grafana/mimir/pkg/querier/worker	5.347s
```

--- a/pkg/querier/worker/scheduler_processor_test.go
+++ b/pkg/querier/worker/scheduler_processor_test.go
@@ -16,7 +16,6 @@ import (
 
 	"github.com/go-kit/log"
 	"github.com/gogo/status"
-	"github.com/grafana/dskit/concurrency"
 	"github.com/grafana/dskit/flagext"
 	"github.com/grafana/dskit/grpcclient"
 	"github.com/grafana/dskit/httpgrpc"
@@ -33,6 +32,7 @@ import (
 	"github.com/grafana/mimir/pkg/querier/stats"
 	querier_stats "github.com/grafana/mimir/pkg/querier/stats"
 	"github.com/grafana/mimir/pkg/scheduler/schedulerpb"
+	"github.com/grafana/mimir/pkg/util/test"
 )
 
 func TestSchedulerProcessor_processQueriesOnSingleStream(t *testing.T) {
@@ -251,8 +251,13 @@ func TestSchedulerProcessor_processQueriesOnSingleStream(t *testing.T) {
 		}, 2*time.Second, 10*time.Millisecond, "expected frontend to be informed of query result exactly once")
 
 		// We expect Send() to be called twice: first to send the querier ID to scheduler
-		// and then to send the query result.
-		loopClient.AssertNumberOfCalls(t, "Send", 2)
+		// and then to send the query result. However, there's no guarantee that the 2nd
+		// Send() has already been called when we reach this point because this test is mocking
+		// several components and there's no real coordination between them, so we poll the assertion.
+		require.EventuallyWithT(t, func(collect *assert.CollectT) {
+			loopClient.AssertNumberOfCalls(test.NewCollectWithLogf(collect), "Send", 2)
+		}, 2*time.Second, 10*time.Millisecond)
+
 		loopClient.AssertCalled(t, "Send", &schedulerpb.QuerierToScheduler{QuerierID: "test-querier-id"})
 	})
 }

--- a/pkg/querier/worker/scheduler_processor_test.go
+++ b/pkg/querier/worker/scheduler_processor_test.go
@@ -16,6 +16,7 @@ import (
 
 	"github.com/go-kit/log"
 	"github.com/gogo/status"
+	"github.com/grafana/dskit/concurrency"
 	"github.com/grafana/dskit/flagext"
 	"github.com/grafana/dskit/grpcclient"
 	"github.com/grafana/dskit/httpgrpc"

--- a/pkg/ruler/compat_test.go
+++ b/pkg/ruler/compat_test.go
@@ -529,7 +529,7 @@ func TestDefaultManagerFactory_CorrectQueryableUsed(t *testing.T) {
 
 			// Ensure the result has been written.
 			require.EventuallyWithT(t, func(collect *assert.CollectT) {
-				pusher.AssertCalled(&collectWithLogf{collect}, "Push", mock.Anything, mock.Anything)
+				pusher.AssertCalled(test.NewCollectWithLogf(collect), "Push", mock.Anything, mock.Anything)
 			}, 5*time.Second, 100*time.Millisecond)
 
 			manager.Stop()
@@ -593,7 +593,7 @@ func TestDefaultManagerFactory_ShouldNotWriteRecordingRuleResultsWhenDisabled(t 
 			if writeEnabled {
 				// Ensure the result has been written.
 				require.EventuallyWithT(t, func(collect *assert.CollectT) {
-					pusher.AssertCalled(&collectWithLogf{collect}, "Push", mock.Anything, mock.Anything)
+					pusher.AssertCalled(test.NewCollectWithLogf(collect), "Push", mock.Anything, mock.Anything)
 				}, 5*time.Second, 100*time.Millisecond)
 			} else {
 				// Ensure no write occurred within a reasonable amount of time.
@@ -846,9 +846,3 @@ func mustStatusWithDetails(code codes.Code, cause mimirpb.ErrorCause) *status.St
 	}
 	return s
 }
-
-type collectWithLogf struct {
-	*assert.CollectT
-}
-
-func (c *collectWithLogf) Logf(_ string, _ ...interface{}) {}

--- a/pkg/util/test/collect.go
+++ b/pkg/util/test/collect.go
@@ -1,0 +1,17 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package test
+
+import "github.com/stretchr/testify/assert"
+
+type CollectWithLogf struct {
+	*assert.CollectT
+}
+
+func NewCollectWithLogf(collectT *assert.CollectT) *CollectWithLogf {
+	return &CollectWithLogf{
+		CollectT: collectT,
+	}
+}
+
+func (c *CollectWithLogf) Logf(_ string, _ ...interface{}) {}

